lib/mayaUsd/ufe/UsdUndoAttributesCommands.h

```diff
@@ -0,0 +1,104 @@
+//
+// Copyright 2022 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#pragma once
+
+#include <mayaUsd/base/api.h>
+#include <mayaUsd/ufe/UsdAttribute.h>
+#include <mayaUsd/ufe/UsdSceneItem.h>
+#include <mayaUsd/undo/UsdUndoableItem.h>
+
+#include <pxr/usd/usd/prim.h>
+
+#include <ufe/attributes.h>
+#ifdef UFE_V4_FEATURES_AVAILABLE
+#if (UFE_PREVIEW_VERSION_NUM >= 4008)
+#include <ufe/nodeDef.h>
+#endif
+#endif
+
+#include <unordered_map>
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+//! \brief Implementation of AddAttributeCommand
+class UsdAddAttributeCommand : public Ufe::AddAttributeCommand
```

Comment:
Any way you can derive from `UsdUndoableCommand<Ufe::AddAttributeCommand>`?  Would save you a bit of code.

Reply:
Not currently. The base class supported by UsdUndoableCommand are required to have a constructor that takes a Ufe::Path. This class was made to wrap commands that are based on BaseUndoableCommand, while my two commands are based on UndoableCommand.

I could go in Ufe and change the Command base classes, but that is actually not trivial.

Reply:
No, let's not change the UFE code, I think it's fine as it is.  Clearly we need a version of UsdUndoableCommand that doesn't have the constructor requirement, since we've already crafted one here:
https://github.com/Autodesk/maya-usd/blob/bc228b4f03216cbe47daf095adaf334ed38e1810/lib/mayaUsd/ufe/UsdAttribute.cpp#L299
We should factor that out, give it a new name, perhaps place it in the same file as the existing UsdUndoableCommand, and make it available for all.  Hey, perhaps someone with excellent template meta-programming skills could manage to re-work the existing UsdUndoableCommand template to call the Cmd base class constructor with a Ufe::Path if one is provided, and call the default constructor if no Ufe::Path is provided...

Reply:
Sadly my metaprogramming skills are not sufficient to select constructors from a base class. This looks like a job for the [Reflection TS](https://en.cppreference.com/w/cpp/experimental/reflect) which seems to have been [recently accepted](https://www.iso.org/standard/76425.html) and  allows browsing the templated base class for constructors to add to the derived class.